capa/features/extractors/dotnet/helpers.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+from typing import Any, Dict, Tuple, Iterator
+from itertools import chain
+
+import dnfile
+from dncil.cil.body import CilMethodBody
+from dncil.cil.error import MethodBodyFormatError
+from dncil.clr.token import Token, StringToken, InvalidToken
+from dncil.cil.body.reader import CilMethodBodyReaderBase
+
+# key indexes to dotnet metadata tables
+DOTNET_META_TABLES_BY_INDEX = {table.value: table.name for table in dnfile.enums.MetadataTables}
+
+
+class DnfileMethodBodyReader(CilMethodBodyReaderBase):
+    def __init__(self, pe: dnfile.dnPE, row: dnfile.mdtable.MethodDefRow):
+        self.pe: dnfile.dnPE = pe
+        self.offset: int = self.pe.get_offset_from_rva(row.Rva)
```

Comment:
we use file offset to keep track of our location in the file; this means file offset is used to calculate instruction offsets which means we emit file offsets for function scope features. see comment above about emitting dotnet tokens.

Reply:
probably useful to add that as a comment/documentation to the source here

Reply:
will be addressed in #981